File: backend/app/schemas/schedule.py
from pydantic import BaseModel, Field
from datetime import date, time, datetime
from typing import Optional

# スケジュール作成（全タイプ統合）
class ScheduleCreate(BaseModel):
    """
    スケジュール作成リクエスト
    """
    schedule_type: str = Field(..., description="return, meal, car, event のいずれか")
    title: Optional[str] = None
    description: Optional[str] = None
    date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    
    # 食事用（schedule_type="meal" の時のみ使用）
    breakfast: Optional[bool] = None
    lunch: Optional[bool] = None
    dinner: Optional[bool] = None
    
    # 車用（schedule_type="car" の時のみ使用）
    car_name: Optional[str] = None 

# ユーザー情報（簡易版）
class ScheduleUserResponse(BaseModel):
    """
    スケジュールに含まれるユーザー情報
    """
    id: int
    username: str
    full_name: Optional[str] = None
    
    class Config:
        from_attributes = True

# スケジュールのレスポンス
class ScheduleResponse(BaseModel):
    """
    スケジュール情報のレスポンス
    """
    id: int
    user_id: int
    family_id: int
    schedule_type: str
    title: Optional[str] = None
    description: Optional[str] = None
    date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    
    # 食事情報
    breakfast: Optional[bool] = None
    lunch: Optional[bool] = None
    dinner: Optional[bool] = None
    
    # 車情報
    car_name: Optional[str] = None
    
    # ユーザー情報
    user: ScheduleUserResponse
    
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True

File: frontend/src/services/schedule.ts
import api from './api';
import type { Schedule } from '../types';

// スケジュール作成用の型
interface ScheduleCreateRequest {
  schedule_type: 'return' | 'meal' | 'car' | 'event';
  title?: string;
  description?: string;
  date: string;  // YYYY-MM-DD 形式
  start_time?: string;  // HH:MM:SS 形式
  end_time?: string;
  breakfast?: boolean;
  lunch?: boolean;
  dinner?: boolean;
  car_name?: string;
}

// スケジュールを作成
export const createSchedule = async (data: ScheduleCreateRequest): Promise<Schedule> => {
  const response = await api.post<Schedule>('/schedules/', data);
  return response.data;
};

// 家族全員のスケジュールを取得
export const getSchedules = async (
  scheduleType?: string,
  startDate?: string,
  endDate?: string
): Promise<Schedule[]> => {
  const params = new URLSearchParams();
  
  if (scheduleType) params.append('schedule_type', scheduleType);
  if (startDate) params.append('start_date', startDate);
  if (endDate) params.append('end_date', endDate);
  
  const response = await api.get<Schedule[]>(`/schedules/?${params.toString()}`);
  return response.data;
};

// 自分のスケジュールのみ取得
export const getMySchedules = async (scheduleType?: string): Promise<Schedule[]> => {
  const params = new URLSearchParams();
  if (scheduleType) params.append('schedule_type', scheduleType);
  
  const response = await api.get<Schedule[]>(`/schedules/me?${params.toString()}`);
  return response.data;
};

// 特定のスケジュールを取得
export const getSchedule = async (scheduleId: number): Promise<Schedule> => {
  const response = await api.get<Schedule>(`/schedules/${scheduleId}`);
  return response.data;
};

// スケジュールを更新
export const updateSchedule = async (
  scheduleId: number,
  data: ScheduleCreateRequest
): Promise<Schedule> => {
  const response = await api.put<Schedule>(`/schedules/${scheduleId}`, data);
  return response.data;
};

// スケジュールを削除
export const deleteSchedule = async (scheduleId: number): Promise<void> => {
  await api.delete(`/schedules/${scheduleId}`);
};